common/logger/src/struct_log.rs

```diff
@@ -426,25 +426,84 @@ impl StructLogSink for UDPStructLog {
     }
 }
 
-struct UDPStructLogThread {
+struct TCPStructLogThread {
     receiver: Receiver<StructuredLogEntry>,
-    socket: UdpSocket,
-    udp_address: String,
+    address: SocketAddr,
 }
 
-impl UDPStructLogThread {
-    pub fn run(self) {
-        for entry in self.receiver {
+impl TCPStructLogThread {
+    fn new(receiver: Receiver<StructuredLogEntry>, address: &str) -> TCPStructLogThread {
+        let address = SocketAddr::from_str(address).unwrap();
+        TCPStructLogThread { receiver, address }
+    }
+
+    // Continually iterate over requests unless writing fails.
+    fn process_requests(&self, stream: &mut TcpStream) {
+        for entry in &self.receiver {
+            // Parse string, skipping over anything that can't be parsed
             let json_string = match entry.to_json_string() {
                 Ok(json_string) => json_string,
                 Err(_) => continue,
             };
-            if let Err(e) = self
-                .socket
-                .send_to(json_string.as_bytes(), self.udp_address.clone())
-            {
-                // do not break on error, move on to the next log message
-                println!("[Logging] Error while sending data to socket: {}", e);
+
+            // If we fail to write, exit out and create a new stream
+            if let Err(e) = stream.write(json_string.as_bytes()) {
+                println!(
+                    "[Logging] Error while sending data to logstash({}): {}",
+                    self.address, e
+                );
+
+                // Attempt to clear the buffer and send whatever we can
+                if let Err(e) = stream.flush() {
+                    println!("[Logging] Failed to flush rest of buffer: {}", e);
+                }
+
+                // Start over stream
+                return;
```

Comment:
we'll lose the log here, no?

Reply:
should we be sending this to our other logger?

Reply:
It now sends metrics that says it crashed :)